ambry-store/src/main/java/com/github/ambry/store/IndexSegment.java

```diff
@@ -828,10 +828,23 @@ void writeIndexSegmentToFile(Offset safeEndPoint) throws FileNotFoundException,
    * @throws StoreException if there are problems with the index
    */
   void seal() throws StoreException {
-    sealed.set(true);
-    map();
-    // we should be fine reading bloom filter here without synchronization as the index is read only
-    persistBloomFilter();
+    rwLock.writeLock().lock();
+    try {
+      sealed.set(true);
+      map();
+      // we should be fine reading bloom filter here without synchronization as the index is read only
+      persistBloomFilter();
+    } catch (StoreException e) {
+      // If the in-memory map is already set to null, it means we are here because we failed writing the bloom filter
+      // to disk. The index segment can still function with bloom filter file being missing, so don't know anything
+      // in this case.
+      if (index != null) {
+        sealed.set(false);
+      }
+      throw e;
+    } finally {
+      rwLock.writeLock().unlock();
+    }
```

Comment:
There are a lot of places where sealed.get() is called and it seems like those could exhibit similar issues (albeit less frequently called methods than find). We may want to make a helper method to check for sealed state that always acquires the read lock